File: web/src/modules/catalogo/scale.test.ts
import 'fake-indexeddb/auto'
import { describe, expect, it } from 'vitest'
import { bootstrapScaleRules, fetchScaleRules, parseScaleCode } from './scale'
import { PosDatabase, getScaleRules } from '../../pos/adapters/dexie'

describe('regla de balanza', () => {
  const rule = { code: 'MEAT', prefix: '99', weightDigits: 3, priceDigits: 5, priority: 1 }
  it('produce el mismo formato offline', () => expect(parseScaleCode('9912301234', [rule])).toEqual({ itemCode: 'MEAT', quantity: '123', price: '01234' }))
  it('rechaza precio cero', () => expect(() => parseScaleCode('9912300000', [rule])).toThrow())
  it('ignora reglas malformadas', () => expect(parseScaleCode('9912301234', [{ ...rule, weightDigits: 0 }, rule])).toEqual({ itemCode: 'MEAT', quantity: '123', price: '01234' }))
  it('descarga reglas V4 por empresa', async () => {
    const fetcher = async () => new Response(JSON.stringify({ rules: [{ prefix: '99', length: 10, product_start: 3, product_length: 3, value_start: 6, value_length: 5, decimals: 2, content: 'PRECIO' }] }), { status: 200 })
    await expect(fetchScaleRules('http://central', '4', fetcher)).resolves.toHaveLength(1)
  })
  it('descarga y persiste reglas para modo offline', async () => {
    const db = new PosDatabase(`scale-bootstrap-${crypto.randomUUID()}`)
    const fetcher = async () => new Response(JSON.stringify({ rules: [{ prefix: '98', length: 10, product_start: 3, product_length: 3, value_start: 6, value_length: 5, decimals: 2, content: 'PESO' }] }), { status: 200 })
    await bootstrapScaleRules('http://central', '4', db, fetcher)
    expect(await getScaleRules('4', db)).toHaveLength(1)
    await db.delete()
  })
  it('rechaza reglas V4 con posiciones inválidas', async () => {
    const fetcher = async () => new Response(JSON.stringify({ rules: [{ prefix: '99', length: 4, product_start: 3, product_length: 3, value_start: 1, value_length: 1, decimals: 2, content: 'PRECIO' }] }), { status: 200 })
    await expect(fetchScaleRules('http://central', '4', fetcher)).rejects.toThrow('INVALID_SCALE_RULES')
  })
  it('rechaza campos numéricos con tipo incorrecto', async () => {
    const fetcher = async () => new Response(JSON.stringify({ rules: [{ prefix: '99', length: '10', product_start: 3, product_length: 3, value_start: 6, value_length: 5, decimals: 2, content: 'PESO' }] }), { status: 200 })
    await expect(fetchScaleRules('http://central', '4', fetcher)).rejects.toThrow('INVALID_SCALE_RULES')
  })
  it('usa caché cuando el servidor no está disponible', async () => {
    const db = new PosDatabase(`scale-offline-${crypto.randomUUID()}`)
    const rules = [{ prefix: '97', length: 10, product_start: 3, product_length: 3, value_start: 6, value_length: 5, decimals: 2, content: 'PESO' as const }]
    const firstFetcher = async () => new Response(JSON.stringify({ rules }), { status: 200 })
    await bootstrapScaleRules('http://central', '4', db, firstFetcher)
    const offlineFetcher = async () => { throw new Error('offline') }
    await expect(bootstrapScaleRules('http://central', '4', db, offlineFetcher)).resolves.toEqual(rules)
    await db.delete()
  })
})

File: internal/modules/catalogo/api/scale_test.go
package api

import "testing"

func TestScaleCodeExactRuleWins(t *testing.T) {
	got, err := ParseScaleCode("1234567890", []ScaleRule{{Code: "1234567890", Prefix: "12", WeightDigits: 3, PriceDigits: 5, Priority: 1}})
	if err != nil || got.ItemCode != "1234567890" {
		t.Fatalf("got=%#v err=%v", got, err)
	}
}
func TestScaleCodeRejectsZeroPrice(t *testing.T) {
	if _, err := ParseScaleCode("9912300000", []ScaleRule{{Code: "MEAT", Prefix: "99", WeightDigits: 3, PriceDigits: 5}}); err != ErrInvalidScaleValue {
		t.Fatal("zero price accepted")
	}
}

func TestScaleCodePreservesFixedWidthDigits(t *testing.T) {
	got, err := ParseScaleCode("9912301234", []ScaleRule{{Code: "MEAT", Prefix: "99", WeightDigits: 3, PriceDigits: 5}})
	if err != nil {
		t.Fatal(err)
	}
	if got.Quantity != "123" || got.Price != "01234" {
		t.Fatalf("expected fixed-width values, got %#v", got)
	}
}

func TestScaleCodeIgnoresMalformedRules(t *testing.T) {
	if _, err := ParseScaleCode("123456", []ScaleRule{{Code: "BAD", Prefix: "12", WeightDigits: -1, PriceDigits: 5}}); err != ErrInvalidScaleValue {
		t.Fatalf("expected malformed rule rejection, got %v", err)
	}
}

func TestParseV4ScaleCodeUsesConfiguredPositions(t *testing.T) {
	rule := V4ScaleRule{Prefix: "99", Length: 11, ProductStart: 3, ProductLength: 3, ValueStart: 6, ValueLength: 5, Decimals: 2, Content: "PRECIO"}
	got, err := ParseV4ScaleCode("99123001234", []V4ScaleRule{rule})
	if err != nil || got.ItemCode != "123" || got.Price != "00123" {
		t.Fatalf("got=%#v err=%v", got, err)
	}
}

func TestParseV4ScaleCodeRejectsSignedFields(t *testing.T) {
	rule := V4ScaleRule{Prefix: "99", Length: 6, ProductStart: 3, ProductLength: 2, ValueStart: 5, ValueLength: 2, Decimals: 2, Content: "PRECIO"}
	if _, err := ParseV4ScaleCode("99+1+2", []V4ScaleRule{rule}); err != ErrInvalidScaleValue {
		t.Fatalf("signed fixed-width field accepted: %v", err)
	}
}
